src/vos/vos_obj.c

```diff
@@ -24,55 +24,114 @@ D_CASSERT((uint32_t)VOS_VIS_FLAG_VISIBLE == (uint32_t)EVT_VISIBLE);
 D_CASSERT((uint32_t)VOS_VIS_FLAG_PARTIAL == (uint32_t)EVT_PARTIAL);
 D_CASSERT((uint32_t)VOS_VIS_FLAG_LAST == (uint32_t)EVT_LAST);
 
+struct vos_key_info {
+	umem_off_t		*ki_known_key;
+	struct vos_object	*ki_obj;
+	bool			 ki_non_empty;
+	bool			 ki_has_uncommitted;
+	const void		*ki_first;
+	daos_handle_t		 ki_toh;
+};
+
 /** This callback is invoked only if the tree is not empty */
 static int
 empty_tree_check(daos_handle_t ih, vos_iter_entry_t *entry,
 		 vos_iter_type_t type, vos_iter_param_t *param, void *cb_arg,
 		 unsigned int *acts)
 {
-	bool	*empty = cb_arg;
+	struct vos_rec_bundle	 rbund = {0};
+	d_iov_t			 val_iov;
+	struct umem_instance	*umm;
+	struct vos_key_info	*kinfo = cb_arg;
+	int			 rc;
+
+	if (kinfo->ki_first == entry->ie_key.iov_buf)
+		return 1; /** We've seen this one before */
 
-	*empty = false;
+	/** Save the first thing we see so we can stop iteration early
+	 *  if we see it again on 2nd pass.
+	 */
+	if (kinfo->ki_first == NULL)
+		kinfo->ki_first = entry->ie_key.iov_buf;
+
+	if (entry->ie_vis_flags == VOS_IT_UNCOMMITTED) {
+		kinfo->ki_has_uncommitted = true;
+		return 0;
+	}
+
+	d_iov_set(&val_iov, &rbund, sizeof(rbund));
+	rc = dbtree_fetch(kinfo->ki_toh, BTR_PROBE_EQ, DAOS_INTENT_UPDATE, &entry->ie_key, NULL,
```

Comment:
Use dbtree_iter_fetch() could save a tree lookup.

Reply:
I will look into it.